### ISSUE TITLE
Window Borders

### ISSUE THREAD

**Issue by dimitrik-fr** [cosmos72/twin#22]
Hello Max,

from the .twinrc example I can see there is a possibility to change window borders via:

Border "*" Active (
    "ÉÍ»"
    "º º"
    "ÈÍ¼"
 )

what kind of characters are expected here ?.. -- everything I'm trying is just giving me kind of "hieroglyphs" printed around window ;-))

even the following "simple" one:

Border "*" Active (
    "+=+"
    "| |"
    "+=+"
) 

Thank you for your help !

Rgds,
-Dimitri

**Comment by cosmos72** [comment]
Hello Dimitri,

you are supposed to be use codepage 437 (i.e. vga) bytes for borders.
It's very old code - I checked it, it seems bugged now.

A better, modern choice would have been to always use UTF-8 inside .twinrc

I will have a look when I have a little time.

**Comment by dimitrik-fr** [comment]
Hello Max,

but what is so special with window "borders" characters vs for ex. "button" characters ?..
for my buttons I can use the following without any problem:
#  
    Button 0 "\311x" Left
    Button 1 "\376\273" Right
    Button 2 "\315\371" Right
#
(I found it much more pleasant for eyes than default []/></etc..)

any reason why the same is not possible for window borders ?..
(I'm not going out of 437 codepage (vga), right ?)

also, as I've mention buttons.. -- any special reason why exactly 2 characters required to label a button? (and 1 or 3 are not allowed ?)..

Thanks again for all your help !!!

Rgds,
-Dimitri

**Comment by lighth7015** [comment]
Can confirm this also messes up on PuTTy.

**Comment by cosmos72** [comment]
Maybe there is some issue with customizing window borders?
I wouldn't say it's high priority, but I will try to look at the situation.

Also, being restricted to codepage 437 (VGA) is clearly an obsolete constraint nowadays... Most of the remaining sources in twin use 2-byte subset of Unicode i.e UCS-2, so it's definitely possible to do better (I also started upgrading to UTF-32, but never finished the work). Contributions are welcome :)

Why exactly 2 characters are required to label a button: it was a design decision copied from TurboVision in 1992 (really) and it is still coded in that way. To be honest, I don't remember if it can be changed or it's hardcoded. I will have a look.

**Comment by lighth7015** [comment]
So how would someone go about changing the way window borders are drawn/rendered? Would that be in, e.g., the `tty` driver?

**Comment by cosmos72** [comment]
The function `WMFindBorderWindow` in `server/wm.c` defines what's in a window border, and how it reacts to mouse events.
To change the look and behaviour of window borders, you need to modify that.

**Comment by lighth7015** [comment]
Thanks! :D

**Comment by dimitrik-fr** [comment]
Would be also great to make border colors configurable. I understand that the code is no more actively developed, but it may still get a new life again due it's lightweight + progressive adoption of Cloud (and Cloud means "remote access", where low latency and low network traffic are key points for comfortable use of command line ;-))

For my part to change border colors to something more "quiet" in grey scale I could only hack little bit the source code to obtain the following :

<img width="1074" alt="68AD7963-A471-49C4-AC95-8A58F87F8B02" src="https://user-images.githubusercontent.com/23528934/56092453-8503eb80-5ebc-11e9-8387-c1a1c3d33aef.png">

looks this way well in iTerm2 on Mac and in Konsole on Linux (no idea why other "Terminal" apps would show it more "ugly", but as long as there is at least one solution which works for me, I'm fine ;-))

All in all, my sincere KUDOS to Max for this great tool !!! -- may only regret to not find TWIN many years before, while now cannot imagine to live without ;-))

Rgds,
-Dimitri

**Comment by lighth7015** [comment]
@dimitrik-fr yeah, I was actually looking at giving it avery similar to how VBDOS looked back on MS-DOS, haha.

**Comment by lighth7015** [comment]
Honestly the server should be overhauled.

**Comment by DJGoossens** [comment]
I am trying to use twin in Linux framebuffer. It works perfectly well, thank you, except the active window has a row of ugly diamonds around it. Menus and inactive windows are fine. I have tried playing with .twinrc, referring to the 437 code table, and I can see what it should be doing but it does not seem to be doing it.  I cannot seem to get it to work. I am guessing the problem lies with where the double-line borders are in the code table of the font I am using. Can you suggest a framebuffer font that should work? I have tried quite a few... any assistance would be welcome...

**Comment by cosmos72** [comment]
Confirmed, I have the same issue on  Linux framebuffer. I will investigate it

**Comment by DJGoossens** [comment]
Thanks very much.
